## MIIA-4203 MODELOS AVANZADOS PARA ANÁLISIS DE DATOS II


# Introducción a Sistemas de Recomendación

## Actividad 10


### Profesor: Camilo Franco (c.franco31@uniandes.edu.co)

## Actividad en grupos
### Nombres:

### Daniel León - David Tavera - Grupo 14

**Instrucciones:** Por favor escriba los nombres de los integrantes de su grupo. Esta actividad debe ser entregada a más tardar dentro de 8 días, con la respuesta para los ejercicios y preguntas en cada numeral.

En este cuadernos estudiaremos una primera introducción a los sistemas de recomendación, examinando un modelo de recomendación en base a la popularidad, sin ningún grado de personalización. En futuras actividades exploraremos otros sistemas basados en contenido o en filtrado colaborativo, los cuales toman en cuenta las preferencias de los usuarios. 

Los sistemas de recomendación están entre las aplicaciones más populares de la ciencia de datos actualmente. El marco que ofrecen estos sistemas permite representar modelos de decisión junto con métodos estadístico/computacionales para **aprender y estimar** las preferencias de los usuarios. 


## Introducción

En esta introducción vamos a construir un modelo simple basado en popularidad. Aprenderemos nociones que nos permitirán construir modelos más complejos de recomendación. 

En este cuaderno vamos a trabajar con una base de datos de películas IMDB (https://www.imdb.com/) 

En términos generales, los sistemas de recomendación pueden clasificarse en 3 tipos:

- **Recomendadores simples:** genera recomendaciones para cada usuario según la popularidad del ítem o producto. Por ejemplo, proponer películas o un género en base a su popularidad o su crítica, donde una buena crítica aproxima una mayor probabilidad de que la película sea apreciada por una audiencia "promedio".
- **Recomendadores basados en contenido:** sugiere elementos similares en función de un elemento en particular, como por ejemplo el género, la sinopsis o el reparto (actores, etc). De esta manera, si el usuario tiene unas preferencias específicas sobre un ítem/artículo en particular, también podría tener preferencia por un ítem *similar*.
- **Recomendadores por filtrado colaborativo:** estos sistemas generan recomendaciones a partir de similitudes entre usuarios, prediciendo la calificación o preferencia de un usuario sobre un ítem, o un grupo de atributos de dicho ítem, según las calificaciones y preferencias observadas de otros usuarios. 

Por último se puede considerar también un cuarto tipo de recomendadores, que son los de tipo **Híbrido**. Estos hacen referencia al uso simultaneo de las distintas metodologias antes señaladas y de múltiples fuentes de información o tecnologías específicas, como pueden ser las redes sociales o fuentes de información geográfica, demográfica, etc.

Para profundizar en los sistemas de recomendación se pueden consultar las siguientes fuentes:
- Ch. Aggarwal. Recommender Systems. Springer, 2016.
- J. Leskovec, A. Rajaraman, J. Ullman. Mining of Massive Datasets. Cambridge University Press (http://www.mmds.org/)


## 1. Sistema de recomendación simple

Los recomendadores simples generan recomendaciones basadas en su popularidad, de acuerdo con una *preferencia promedio*. 

La pregunta es,

**¿cómo medimos la popularidad de un producto/ítem?**

Una opción es la de elegir una métrica determinada para puntuar las películas, y en base a esa métrica generar un valor de preferencia por el ítem. De esta manera es posible consturir un *orden lineal* o un *ranking estricto* con las mejores películas de la base de datos IMDB.

Pasos a seguir:
- Elegir la métrica con la que se va a calificar cada película. Este punto también se puede entender como un modelo para estimar las preferencias por una película.
- Calcular el valor de preferencia sobre cada película.
- Ordenar las películas según su valor o puntuación y obtener una lista con las mejores k peliculas.

Manos a la obra, primero carguemos el conjunto de datos como un data frame en la bilbioteca Pandas:

In [1]:
# Importamos la biblioteca Pandas
import pandas as pd

# Cargamos los datos de peliculas de la base de datos IMDB
metadata = pd.read_csv('movies_metadata.csv', low_memory=False)

print(metadata.shape)
      
list(metadata)


(45466, 24)


['adult',
 'belongs_to_collection',
 'budget',
 'genres',
 'homepage',
 'id',
 'imdb_id',
 'original_language',
 'original_title',
 'overview',
 'popularity',
 'poster_path',
 'production_companies',
 'production_countries',
 'release_date',
 'revenue',
 'runtime',
 'spoken_languages',
 'status',
 'tagline',
 'title',
 'video',
 'vote_average',
 'vote_count']

In [2]:
# Así se ven los datos
metadata.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   

     original_title                                           overview  ...  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1           Jumanji  When siblings Judy and Peter discover an encha...  ...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...  ...   

  release_date      revenue runtime  \
0   1995-10-30  373554033.0    81.0   
1   1995-12-15  262797249.0   104.0   
2   1995-12-22          0.0   101.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average vote_count  
0          7.7     5415.0  
1          6.9     2413.0  
2          6.5       92.0  

[3 rows x 24 columns]

In [3]:
metadata[['original_title', 'popularity']][1:10]

original_title popularity
1                      Jumanji  17.015539
2             Grumpier Old Men    11.7129
3            Waiting to Exhale   3.859495
4  Father of the Bride Part II   8.387519
5                         Heat  17.924927
6                      Sabrina   6.677277
7                 Tom and Huck   2.561161
8                 Sudden Death    5.23158
9                    GoldenEye  14.686036

### Rating vs. Popularidad

Con esta información vamos a construir un sistema que recomiende películas basados directamente en su *rating*, o **voto promedio**. Sin embargo, es cierto que hay otro atributo relevante como lo es el de su **popularidad**. Por ejemplo, si un solo usuario califica con 10 una pelicula, pero 5000 usuarios le dan en promedio un 8.3, solo en base al rating la primera película sería considerada mejor que la segunda.


Bajo una perspectiva de *los grandes números*, la valoración o voto promedio de una película tiende a converger hacia un valor que refleje la calidad de la película a medida que el número de votos aumenta. Con poca información, o pocos votos, es más difícil estimar la calidad promedio de la película.  


Entonces, 

**¿qué tipo de métrica nos sirve para evaluar apropiadamente la preferencia por una película?** 

Tengamos en cuenta que en base a esta métrica se tendrán mejores o peores recomendaciones!


#### Respuesta

Podríamos considerar en utilizar una métrica calculada a partir de los **votos promedios**, **cantidad de votos** y **popularidad**, que nos permita ponderar mejor la calificación de la película y evitar sesgar las recomendación al seleccionar películas con un rating alto pero poca popularidad o conteo de votos.

### Ejercicio 1.1

Examine los datos y proponga una métrica que permita estimar la preferencia promedio por una película.

#### Podríamos calcular el rating ponderado a partir de dividir la cantidad de votos de una película sobre el promedio de votos recibidos en las distintas películas de la base, multiplicado por el rating logrado por la película. A continuación la formula del cálculo.

$$
\mu_i  = \left( \frac{v_i}{v_{prom}} \right) R_i 
$$

#### De esta manera, la función de calculo del rating ponderado y la selección del top de películas bajo esta metrica serán calculadas a continuación.

In [4]:
pelisPon = metadata[['original_title', 'homepage', 'vote_count', 'vote_average']]

# Calificacion de peliculas basada en la ponderacion de su rating 
def rating_ponderado_prom(x):
    m = x['vote_average'].mean()
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) 

import warnings
warnings.filterwarnings("ignore", category=Warning)

# si tiene al menos mil votos, incluimos la pelicula
vmin = 1000
pelisRF = pelisPon.copy().loc[pelisPon['vote_count'] >= vmin]
pelisRF['score'] = rating_ponderado_prom(pelisRF)

# ordenamos las peliculas de acuerdo con el score
pelisRF = pelisRF.sort_values('score', ascending=False)

# Mostramos el top15 de las peliculas
pelisRF[['original_title', 'homepage', 'vote_count', 'vote_average', 'score']].head(15)

original_title  \
314           The Shawshank Redemption   
834                      The Godfather   
40251                            君の名は。   
12481                  The Dark Knight   
2843                        Fight Club   
292                       Pulp Fiction   
522                   Schindler's List   
23673                         Whiplash   
5481                          千と千尋の神隠し   
2211                   La vita è bella   
1178            The Godfather: Part II   
1152   One Flew Over the Cuckoo's Nest   
1176                            Psycho   
1184       Once Upon a Time in America   
351                       Forrest Gump   

                                              homepage  vote_count  \
314                                                NaN      8358.0   
834                       http://www.thegodfather.com/      6024.0   
40251  https://www.funimationfilms.com/movie/yourname/      1030.0   
12481     http://thedarkknight.warnerbros.com/dvdsite/     12269.0   
2843        http://www.foxmovies.com/movies/fight-club      9678.0   
292                                                NaN      8670.0   
522                     http://www.schindlerslist.com/      4436.0   
23673                http://sonyclassics.com/whiplash/      4376.0   
5481            http://movies.disney.com/spirited-away      3968.0   
2211                                               NaN      3643.0   
1178                                               NaN      3418.0   
1152                                               NaN      3001.0   
1176                                               NaN      2405.0   
1184                                               NaN      1104.0   
351                                                NaN      8147.0   

       vote_average     score  
314             8.5  8.493122  
834             8.5  8.490460  
40251           8.5  8.444504  
12481           8.3  8.295423  
2843            8.3  8.294199  
292             8.3  8.293525  
522             8.3  8.287354  
23673           8.3  8.287181  
5481            8.3  8.285865  
2211            8.3  8.284607  
1178            8.3  8.283595  
1152            8.3  8.281321  
1176            8.3  8.276705  
1184            8.3  8.249420  
351             8.2  8.193193

### Ponderación entre rating y popularidad

Una opción puede consistir en ponderar el rating que recibe una película por su número de votos. De esta manera se pueden prevenir situaciones donde una película con un voto de 10 obtenga una mejor calificación que una película con 1000 votos con un promedio de 9.5.  

De esta manera, el voto promedio ponderado $\mu_i$, de la $i$-ésima película lo definimos como:

$$
\mu_i  = \left( \frac{v_i}{v_{max}} \right) R_i 
$$

donde $v_i$ es el número de votos para la $i$-ésima película, $v_{max}$ es el máximo número de votos que recibe la película más popular, y $R$ es el rating promedio de la pelicula.


En la base de datos tenemos los valores de $v_i$ (vote_count) y $R$ (vote_average) para cada pelicula. Además podemos agregar un filtro para que considere las películas con un minimo número de votos $v_{min}$.


Nos quedamos con las columnas necesarias para calcular el rating ponderado más el link de la pagina (en caso de que esté disponible):

In [5]:
pelisR = metadata[['original_title', 'homepage', 'vote_count', 'vote_average']]
pelisR.shape

(45466, 4)

In [6]:
# Calificacion de peliculas basada en la ponderacion de su rating 
def rating_ponderado(x):
    """
    Input:
    x: datos de rating y votacion de las peliculas
    m: minimo numero de votos
    Output:
    rating ponderado
    """
    m = x['vote_average'].max()
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) 

Definimos una nueva respuesta 'score' y calculamos su valor con la funcion rating_ponderado().

A continuación ordenamos las películas basados en el score y ofrecemos el top 15 de películas preferidas:

In [7]:
# si tiene al menos vmin votos, incluimos la pelicula
vmin = 1
pelisRF = pelisR.copy().loc[pelisR['vote_count'] >= vmin]
pelisRF['score'] = rating_ponderado(pelisRF)

# ordenamos las peliculas de acuerdo con el score
pelisRF = pelisRF.sort_values('score', ascending=False)

# Mostramos el top15 de las peliculas
pelisRF[['original_title', 'homepage', 'vote_count', 'vote_average', 'score']].head(15)

original_title  \
10309      Dilwale Dulhania Le Jayenge   
314           The Shawshank Redemption   
834                      The Godfather   
40251                            君の名は。   
39085                     Planet Earth   
12481                  The Dark Knight   
2843                        Fight Club   
292                       Pulp Fiction   
522                   Schindler's List   
23673                         Whiplash   
5481                          千と千尋の神隠し   
2211                   La vita è bella   
1178            The Godfather: Part II   
1152   One Flew Over the Cuckoo's Nest   
1176                            Psycho   

                                              homepage  vote_count  \
10309                                              NaN       661.0   
314                                                NaN      8358.0   
834                       http://www.thegodfather.com/      6024.0   
40251  https://www.funimationfilms.com/movie/yourname/      1030.0   
39085                                              NaN       176.0   
12481     http://thedarkknight.warnerbros.com/dvdsite/     12269.0   
2843        http://www.foxmovies.com/movies/fight-club      9678.0   
292                                                NaN      8670.0   
522                     http://www.schindlerslist.com/      4436.0   
23673                http://sonyclassics.com/whiplash/      4376.0   
5481            http://movies.disney.com/spirited-away      3968.0   
2211                                               NaN      3643.0   
1178                                               NaN      3418.0   
1152                                               NaN      3001.0   
1176                                               NaN      2405.0   

       vote_average     score  
10309           9.1  8.964382  
314             8.5  8.489842  
834             8.5  8.485913  
40251           8.5  8.418269  
39085           8.8  8.326882  
12481           8.3  8.293240  
2843            8.3  8.291433  
292             8.3  8.290438  
522             8.3  8.281332  
23673           8.3  8.281076  
5481            8.3  8.279135  
2211            8.3  8.277279  
1178            8.3  8.275788  
1152            8.3  8.272434  
1176            8.3  8.265631

Examinemos el efecto de cambiar el mínimo numero de votos a considerar:

In [8]:
# si tiene al menos mil votos incluimos la pelicula
vmin = 1000
pelisRF = pelisR.copy().loc[pelisR['vote_count'] >= vmin]
pelisRF['score'] = rating_ponderado(pelisRF)

# ordenamos las peliculas de acuerdo con el score
pelisRF = pelisRF.sort_values('score', ascending=False)

# Mostramos el top15 de las peliculas
pelisRF[['original_title', 'homepage', 'vote_count', 'vote_average', 'score']].head(15)

original_title  \
314           The Shawshank Redemption   
834                      The Godfather   
40251                            君の名は。   
12481                  The Dark Knight   
2843                        Fight Club   
292                       Pulp Fiction   
522                   Schindler's List   
23673                         Whiplash   
5481                          千と千尋の神隠し   
2211                   La vita è bella   
1178            The Godfather: Part II   
1152   One Flew Over the Cuckoo's Nest   
1176                            Psycho   
1184       Once Upon a Time in America   
351                       Forrest Gump   

                                              homepage  vote_count  \
314                                                NaN      8358.0   
834                       http://www.thegodfather.com/      6024.0   
40251  https://www.funimationfilms.com/movie/yourname/      1030.0   
12481     http://thedarkknight.warnerbros.com/dvdsite/     12269.0   
2843        http://www.foxmovies.com/movies/fight-club      9678.0   
292                                                NaN      8670.0   
522                     http://www.schindlerslist.com/      4436.0   
23673                http://sonyclassics.com/whiplash/      4376.0   
5481            http://movies.disney.com/spirited-away      3968.0   
2211                                               NaN      3643.0   
1178                                               NaN      3418.0   
1152                                               NaN      3001.0   
1176                                               NaN      2405.0   
1184                                               NaN      1104.0   
351                                                NaN      8147.0   

       vote_average     score  
314             8.5  8.491364  
834             8.5  8.488023  
40251           8.5  8.430429  
12481           8.3  8.294254  
2843            8.3  8.292717  
292             8.3  8.291871  
522             8.3  8.284126  
23673           8.3  8.283909  
5481            8.3  8.282258  
2211            8.3  8.280679  
1178            8.3  8.279410  
1152            8.3  8.276558  
1176            8.3  8.270769  
1184            8.3  8.236584  
351             8.2  8.191454

Veamos en qué consiste esa pelicula en japonés:

In [9]:
import webbrowser

webbrowser.open(pelisRF['homepage'][40251]) 

True

### Ejercicio 1.2

Examine cómo varían los resultados si tomamos directamente la variable de popularidad.

#### Seleccionamos el top15 de peliculas basadas en su popularidad

##### Filtro de peliculas con al menos un voto

In [10]:
pelisP = metadata[['original_title', 'homepage', 'vote_count', 'vote_average','popularity']]

# si tiene al menos un voto, incluimos la pelicula
vmin = 1
pelisPop = pelisP.copy().loc[pelisR['vote_count'] >= vmin]

# ordenamos las peliculas de acuerdo con a popularidad
pelisPop = pelisPop.sort_values('popularity', ascending=False)

# Mostramos el top15 de las peliculas
pelisPop[['original_title', 'homepage', 'vote_count', 'vote_average', 'popularity']].head(15)

original_title  \
536                     Blade Runner   
3493                        Dinosaur   
10517                        Jarhead   
11252                  The Last Kiss   
18029                    Dream House   
1360                          Hamlet   
12053  4 luni, 3 săptămîni și 2 zile   
12488                     Black Dawn   
25394                      Black Sea   
5214                Hollywood Ending   
4225                  Uncommon Valor   
4294              Viskningar och rop   
14043  The Haunting of Molly Hartley   
20993             The Lords of Salem   
12282           Charlie Wilson's War   

                                                homepage  vote_count  \
536               http://www.warnerbros.com/blade-runner      3833.0   
3493                                                 NaN       563.0   
10517                                                NaN       776.0   
11252                                                NaN       125.0   
18029                                                NaN       401.0   
1360                                                 NaN       118.0   
12053                                                NaN       158.0   
12488                                                NaN        21.0   
25394                                                NaN       410.0   
5214                                                 NaN       114.0   
4225                                                 NaN        54.0   
4294                                                 NaN       118.0   
14043  http://www.mylifetime.com/movies/the-haunting-...        48.0   
20993    http://robzombie.com/movies/the-lords-of-salem/       244.0   
12282                  http://www.charliewilsonswar.net/       342.0   

       vote_average popularity  
536             7.9  96.272374  
3493            6.2   9.998415  
10517           6.6   9.997032  
11252           5.9   9.996541  
18029           5.8   9.995617  
1360            7.3   9.994088  
12053           7.4   9.990406  
12488           3.6   9.989719  
25394           6.2   9.989679  
5214            6.4   9.987317  
4225            6.2   9.987072  
4294            7.8   9.983092  
14043           4.3   9.982123  
20993           5.1   9.981121  
12282           6.5   9.979162

In [11]:
# si tiene al menos mil votos, incluimos la pelicula
vmin = 1000
pelisPop = pelisP.copy().loc[pelisR['vote_count'] >= vmin]

# ordenamos las peliculas de acuerdo con a popularidad
pelisPop = pelisPop.sort_values('popularity', ascending=False)

# Mostramos el top15 de las peliculas
pelisPop[['original_title', 'homepage', 'vote_count', 'vote_average', 'popularity']].head(15)

original_title  \
536                    Blade Runner   
3893                What Women Want   
21162              White House Down   
12356                         Rambo   
37927  Neighbors 2: Sorority Rising   
12368                     In Bruges   
2586                 Wild Wild West   
2556                   Notting Hill   
9542                    The Aviator   
13507         Män som hatar kvinnor   
17358               Horrible Bosses   
227                 Dumb and Dumber   
608                  The Aristocats   
16760                  The Mechanic   
21163               The World's End   

                                                homepage  vote_count  \
536               http://www.warnerbros.com/blade-runner      3833.0   
3893                                                 NaN      1021.0   
21162                                                NaN      1910.0   
12356                          http://www.rambofilm.com/      1009.0   
37927                    http://www.neighbors-movie.com/      1490.0   
12368          http://www.filminfocus.com/film/in_bruges      1414.0   
2586                                                 NaN      1042.0   
2556                        http://www.notting-hill.com/      1309.0   
9542                                                 NaN      1526.0   
13507                       http://dragontattoofilm.com/      1020.0   
17358  http://horriblebossesmovie.warnerbros.com/inde...      2341.0   
227                                                  NaN      1894.0   
608                                                  NaN      1287.0   
16760                   http://www.themechanicmovie.com/      1066.0   
21163                                                NaN      1820.0   

       vote_average popularity  
536             7.9  96.272374  
3893            6.1   9.963079  
21162           6.3   9.920223  
12356           6.4    9.91556  
37927           5.6   9.914008  
12368           7.4   9.895652  
2586            5.1   9.887602  
2556            7.0   9.874043  
9542            7.0    9.86738  
13507           7.4   9.858687  
17358           6.4   9.856955  
227             6.5   9.844558  
608             7.1   9.834047  
16760           6.3   9.832342  
21163           6.7   9.830382

#### Respuesta

#### Podemos observar que la seleccion del top de peliculas cambia considerablemente al tomar la variable popularidad en lugar de la variable calculada del score. Adicionalmente, observamos que la pelicula 'Blade Runner' tiene una popularidad de 96.27, mientras que la segunda pelicula 'What Women Want' tiene un valor de 9.96, por lo cual parece que el valor de Blade Runner es incorrecto.

### Recomendación de peliculas mas populares por genero

In [12]:
import numpy as np
from ast import literal_eval

# trabajamos la informacion por generos
metadata['genres'] = metadata['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

# añadimos la variable del año
metadata['year'] = pd.to_datetime(metadata['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

metadata.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                         genres                              homepage     id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story    862   
1  [Adventure, Fantasy, Family]                                   NaN   8844   
2             [Romance, Comedy]                                   NaN  15602   

     imdb_id original_language    original_title  \
0  tt0114709                en         Toy Story   
1  tt0113497                en           Jumanji   
2  tt0113228                en  Grumpier Old Men   

                                            overview  ...      revenue  \
0  Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1  When siblings Judy and Peter discover an encha...  ...  262797249.0   
2  A family wedding reignites the ancient feud be...  ...          0.0   

  runtime                                   spoken_languages    status  \
0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2   101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average vote_count  year  
0          7.7     5415.0  1995  
1          6.9     2413.0  1995  
2          6.5       92.0  1995  

[3 rows x 25 columns]

Primero nos quedamos con todos los generos:

In [13]:
generos = metadata.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
generos.name = 'genre'
gen_md = metadata.drop('genres', axis=1).join(generos)

gen_md.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   

                               homepage   id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story  862  tt0114709                en   
0  http://toystory.disney.com/toy-story  862  tt0114709                en   
0  http://toystory.disney.com/toy-story  862  tt0114709                en   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   

  popularity  ... runtime                          spoken_languages    status  \
0  21.946943  ...    81.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
0  21.946943  ...    81.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
0  21.946943  ...    81.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

  tagline      title  video vote_average vote_count  year      genre  
0     NaN  Toy Story  False          7.7     5415.0  1995  Animation  
0     NaN  Toy Story  False          7.7     5415.0  1995     Comedy  
0     NaN  Toy Story  False          7.7     5415.0  1995     Family  

[3 rows x 25 columns]

Construimos una funcion para un género particular y que tome en cuenta peliculas con un número vmin de votos:

In [14]:
def rec_gen(genero, vmin):
    df = gen_md[gen_md['genre'] == genero]
    v = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    R = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    m = df['vote_average'].max()
    
    pelisG = df[(df['vote_count'] >= vmin) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'overview', 'homepage']]
    pelisG['vote_count'] = pelisG['vote_count'].astype('int')
    pelisG['vote_average'] = pelisG['vote_average'].astype('int')
    
    pelisG['wr'] = v/m * R
    pelisG = pelisG.sort_values('wr', ascending=False).head(250)
    
    return pelisG



Veamos el Top-15 de recomendaciones en Ciencia Ficción:

In [15]:
k = 15
scifi = rec_gen('Science Fiction', 1000)
scifi.head(15)

title  year  vote_count  vote_average  \
15480                     Inception  2010       14075             8   
22879                  Interstellar  2014       11187             8   
14551                        Avatar  2009       12114             7   
17818                  The Avengers  2012       12000             7   
23753       Guardians of the Galaxy  2014       10014             7   
26553            Mad Max: Fury Road  2015        9629             7   
2458                     The Matrix  1999        9079             7   
12588                      Iron Man  2008        8951             7   
18244              The Hunger Games  2012        9634             6   
26555  Star Wars: The Force Awakens  2015        7993             7   
256                       Star Wars  1977        6778             8   
20830                    Iron Man 3  2013        8951             6   
25084                Jurassic World  2015        8842             6   
26567    Captain America: Civil War  2016        7462             7   
30051                   The Martian  2015        7442             7   

       popularity                                           overview  \
15480   29.108149  Cobb, a skilled thief who commits corporate es...   
22879   32.213481  Interstellar chronicles the adventures of a gr...   
14551  185.070892  In the 22nd century, a paraplegic Marine is di...   
17818   89.887648  When an unexpected enemy emerges and threatens...   
23753   53.291601  Light years from Earth, 26 years after being a...   
26553    29.36178  An apocalyptic story set in the furthest reach...   
2458    33.366332  Set in the 22nd century, The Matrix tells the ...   
12588   22.073099  After being held captive in an Afghan cave, bi...   
18244   20.031667  Every year in the ruins of what was once North...   
26555   31.626013  Thirty years after defeating the Galactic Empi...   
256     42.149697  Princess Leia is captured and held hostage by ...   
20830   23.721243  When Tony Stark's world is torn apart by a for...   
25084   32.790475  Twenty-two years after the events of Jurassic ...   
26567  145.882135  Following the events of Age of Ultron, the col...   
30051    25.62993  During a manned mission to Mars, Astronaut Mar...   

                                                homepage       wr  
15480              http://inceptionmovie.warnerbros.com/  11260.0  
22879                  http://www.interstellarmovie.net/   8949.6  
14551                        http://www.avatarmovie.com/   8479.8  
17818                  http://marvel.com/avengers_movie/   8400.0  
23753                        http://marvel.com/guardians   7009.8  
26553                        http://www.madmaxmovie.com/   6740.3  
2458                    http://www.warnerbros.com/matrix   6355.3  
12588                       http://www.ironmanmovie.com/   6265.7  
18244                   http://www.thehungergames.movie/   5780.4  
26555  http://www.starwars.com/films/star-wars-episod...   5595.1  
256    http://www.starwars.com/films/star-wars-episod...   5422.4  
20830                         http://marvel.com/ironman3   5370.6  
25084                      http://www.jurassicworld.com/   5305.2  
26567           http://marvel.com/captainamericapremiere   5223.4  
30051        http://www.foxmovies.com/movies/the-martian   5209.4

Inception e Interstellar aparecen en las dos primeras posiciones. Personalmente me gusta más Interstellar (si quieres ver un agujero negro, esta película es lo mejor que podrás conseguir), pero reconozco que Inception tiene mucho nivel. Podemos seguir refinando este tipo de recomendaciones prestando atención a los distintos atributos que tenemos disponibles sobre las películas. 


### Ejercicio 1.3

Elija el género de su interés y aplique la métrica propuesta en el Ejercicio 1.1

In [16]:
### Visualizamos los distintos generos disponibles
print(gen_md['genre'].unique())

['Animation' 'Comedy' 'Family' 'Adventure' 'Fantasy' 'Romance' 'Drama'
 'Action' 'Crime' 'Thriller' 'Horror' 'History' 'Science Fiction'
 'Mystery' 'War' 'Foreign' nan 'Music' 'Documentary' 'Western' 'TV Movie'
 'Carousel Productions' 'Vision View Entertainment'
 'Telescene Film Group Productions' 'Aniplex' 'GoHands' 'BROSTA TV'
 'Mardock Scramble Production Committee' 'Sentai Filmworks'
 'Odyssey Media' 'Pulser Productions' 'Rogue State' 'The Cartel']


In [17]:
### Seleccionamos el genero Western, debido a que nos genera interés conocer el listado de peliculas de este género.
western = rec_gen('Western', 1000)
western.head(15)

title  year  vote_count  vote_average  \
20051                   Django Unchained  2012       10297             7   
31865                       The Revenant  2015        6558             7   
28131                  The Hateful Eight  2015        4405             7   
1159      The Good, the Bad and the Ugly  1966        2371             8   
16738                              Rango  2011        2094             6   
16278                          True Grit  2010        1701             7   
21175                    The Lone Ranger  2013        2361             5   
39932              The Magnificent Seven  2016        2358             5   
16492                   Cowboys & Aliens  2011        2054             5   
1166        Once Upon a Time in the West  1968        1160             8   
39663                 Hell or High Water  2016        1304             7   
1221                          Unforgiven  1992        1133             7   
583                   Dances with Wolves  1990        1084             7   
23459  A Million Ways to Die in the West  2014        1494             5   
12042                       3:10 to Yuma  2007        1210             6   

      popularity                                           overview  \
20051  19.785025  With the help of a German bounty hunter, a fre...   
31865  23.501228  In the 1820s, a frontiersman, Hugh Glass, sets...   
28131  20.328822  Bounty hunters seek shelter from a raging bliz...   
1159   16.788787  While the Civil War rages between the Union an...   
16738  10.581925  When Rango, a lost family pet, accidentally wi...   
16278  10.730765  Following the murder of her father by hired ha...   
21175  12.729104  The Texas Rangers chase down a gang of outlaws...   
39932  13.809622  Seven gun men in the old west gradually come t...   
16492  10.958669  A stranger stumbles into the desert town of Ab...   
1166   15.589351  This classic western masterpiece is an epic fi...   
39663  12.565896  A divorced dad and his ex-con brother resort t...   
1221   10.942149  William Munny is a retired, once-ruthless kill...   
583    11.654349  Wounded Civil War soldier, John Dunbar tries t...   
23459   10.38391  As a cowardly farmer begins to fall for the my...   
12042  18.079094  In Arizona in the late 1800's, infamous outlaw...   

                                                homepage      wr  
20051                         http://unchainedmovie.com/  7207.9  
31865       http://www.foxmovies.com/movies/the-revenant  4590.6  
28131                        http://thehatefuleight.com/  3083.5  
1159   http://www.mgm.com/#/our-titles/766/The-Good,-...  1896.8  
16738                         http://www.rangomovie.com/  1256.4  
16278                      http://www.truegritmovie.com/  1190.7  
21175              http://disney.go.com/the-lone-ranger/  1180.5  
39932  http://www.sonypictures.com/movies/themagnific...  1179.0  
16492              http://www.cowboysandaliensmovie.com/  1027.0  
1166                                                 NaN   928.0  
39663                      http://hellorhighwater.movie/   912.8  
1221                                                 NaN   793.1  
583                                                  NaN   758.8  
23459                                                NaN   747.0  
12042                   http://www.310toyumathefilm.com/   726.0

#### Como era de esperarse, 'Django Unchained' y 'The Revenant' son las primeras películas de la lista, personalmente ambas películas son muy buenos referentes del género y en ambas contamos con la actuación de Leonardo Dicaprio.